src/ddscxx/include/org/eclipse/cyclonedds/topic/datatopic.hpp

```diff
@@ -263,7 +282,7 @@ ddsi_serdata_t *serdata_from_sample(
       assert(0);
     }
     d->key_md5_hashed() = msg->key(d->key());
-    d->getT() = *msg;
```

Comment:
Somehow I get the impression that an unnecessary copy step is being done here, the sample pointer is loaned, right?

Reply:
The line you quote here is deleted, but the line that replaces it indeed makes a full copy of a sample. What is important is the context of this piece of code: the serdata_to_sample function. Normally this function is not called by the loaned C++ read/take, which uses read/take_cdr under the hood and just returns the serdata istself. However, in case of a read/take that uses pre-allocated resources I will have to make a copy of the sample into the preallocated resources. It is only in these cases that the serdata_to_sample will be invoked, and in those cases it is conceptually correct.